rust/extractor/src/main.rs

```diff
@@ -1,13 +1,48 @@
 use anyhow::Context;
 use ra_ap_ide_db::line_index::LineIndex;
 use ra_ap_parser::Edition;
+use std::borrow::Cow;
 mod archive;
 mod config;
 pub mod generated;
 mod translate;
 pub mod trap;
 use ra_ap_syntax::ast::SourceFile;
-use ra_ap_syntax::AstNode;
+use ra_ap_syntax::{AstNode, SyntaxError, TextRange, TextSize};
+
+fn from_utf8_lossy(v: &[u8]) -> (Cow<'_, str>, Option<SyntaxError>) {
+    let mut iter = v.utf8_chunks();
+    let (first_valid, first_invalid) = if let Some(chunk) = iter.next() {
+        let valid = chunk.valid();
+        let invalid = chunk.invalid();
+        if invalid.is_empty() {
+            debug_assert_eq!(valid.len(), v.len());
+            return (Cow::Borrowed(valid), None);
+        }
+        (valid, invalid)
+    } else {
+        return (Cow::Borrowed(""), None);
+    };
```

Comment:
Is the special handling of the first chunk essentially an optimisation for the common case?

Reply:
No, it is a way to get the location of the first UTF8 decoding error, if any. If a file is complete garbage we don't want to generate a ton of useless decoding errors, one is enough.

Reply:
Although the `Cow` (copy-on-write) stuff is an optimisation for the common case that the input is entirely valid utf8 and no replacements are needed.